nuget/lib/dependabot/nuget/discovery/discovery_json_reader.rb

```diff
@@ -0,0 +1,94 @@
+# typed: strong
+# frozen_string_literal: true
+
+require "json"
+
+require "dependabot/dependency"
+
+module Dependabot
+  module Nuget
+    class DiscoveryJsonReader
+      extend T::Sig
+
+      DISCOVERY_JSON_PATH = ".dependabot/discovery.json"
+
+      sig { returns(T::Boolean) }
+      private_class_method def self.test_run?
+        ENV["DEPENDABOT_NUGET_TEST_RUN"] == "true"
+      end
+
+      sig { returns(String) }
+      private_class_method def self.temp_directory
+        Dir.tmpdir
+      end
+
+      sig { returns(String) }
+      def self.discovery_file_path
+        File.join(temp_directory, DISCOVERY_JSON_PATH)
+      end
+
+      sig { params(discovery_json: DependencyFile).void }
+      def self.write_discovery_json(discovery_json)
+        throw "Not a test run!" unless test_run?
+
+        @test_discovery_json = T.let(discovery_json, T.nilable(DependencyFile))
+      end
+
+      sig { returns(T.nilable(DependencyFile)) }
+      def self.discovery_json
+        return @test_discovery_json if test_run?
+
+        return unless File.exist?(discovery_file_path)
+
+        DependencyFile.new(
+          name: Pathname.new(discovery_file_path).cleanpath.to_path,
+          directory: temp_directory,
+          type: "file",
+          content: File.read(discovery_file_path)
+        )
+      end
+
+      sig { params(discovery_json: DependencyFile).void }
+      def initialize(discovery_json:)
+        @discovery_json = discovery_json
+      end
+
+      sig { returns(Dependabot::FileParsers::Base::DependencySet) }
+      def dependency_set
+        dependency_set = Dependabot::FileParsers::Base::DependencySet.new
+        return dependency_set unless workspace_discovery
+
+        workspace_result = T.must(workspace_discovery)
+        workspace_result.projects.each do |project|
+          dependency_set += project.dependency_set
+        end
+        if workspace_result.directory_packages_props
+          dependency_set += T.must(workspace_result.directory_packages_props).dependency_set
+        end
+        if workspace_result.dotnet_tools_json
+          dependency_set += T.must(workspace_result.dotnet_tools_json).dependency_set
+        end
+        dependency_set += T.must(workspace_result.global_json).dependency_set if workspace_result.global_json
```

Comment:
It's a bummer that sorbet isn't better at detecting when a nilable type has already been confirmed as non-nil.

Reply:
Yeah, it doesn't seem to track nilable information about members only the object.

Reply:
Here's the documentation on [flow-sensitive typing][1], specifically on its limitations:

> Flow-sensitive type checking only works for local variables, not for values returned from method calls. Why? Sorbet can’t know that if a method is called twice in a row that it returns the same thing each time. Put another way, Sorbet never assumes that a method call is pure.

[1]: https://sorbet.org/docs/flow-sensitive#limitations-of-flow-sensitivity